### PR TITLE
Seperate CAPI server and browser

### PR DESCRIPTION
**Pull request by liywjl**
## What does this change?
To start reducing CAPI size we need to explicitly declare what is send to the browser. By doing this we now diverge between having CAPI browser and CAPI server

## Why?
We send too much data in the `window.guardian` object, in order to reduce the size we need to explicitly state what we are using to avoid sending unnecessary data. CAPI is the first in a series of optimisations we plan on doing on this object.

## Performance
Having run agains master 4x we get an averge of 60 points on master and 70 points on this branch.

- MASTER
Screenshot 2020-03-11 at 14 38 27

- THIS BRANCH
Screenshot 2020-03-11 at 14 38 39

## Link to supporting Trello card
https://trello.com/c/MfrTjOAI/1220-reduce-the-size-of-windowguardian

### CODE REVIEW

File: src/model/window-guardian.ts [hunk 1]
```diff
@@ -66,18 +66,67 @@ const makeWindowGuardianConfig = (
     } as WindowGuardianConfig;
 };
 
+export const makeGuardianBrowserCAPI = (CAPI: CAPIType): CAPIBrowserType => ({
+    config: {
+        isDev: process.env.NODE_ENV !== 'production',
+        ajaxUrl: CAPI.config.ajaxUrl,
+        shortUrlId: CAPI.config.shortUrlId,
+        pageId: CAPI.config.pageId,
+        isPaidContent: !!CAPI.config.isPaidContent,
```
Comment (gtrufitt) [inline src/model/window-guardian.ts]: Can we be explicit here using `===`?

Comment (liywjl) [reply src/model/window-guardian.ts]: `CAPI.config.isPaidContent === true ? true : false` ?

Comment (gtrufitt) [reply src/model/window-guardian.ts]: sorry yeah. Why is it not just  `isPaidContent: CAPI.config.isPaidContent`?

Comment (liywjl) [reply src/model/window-guardian.ts]: I know mine and @oliverlloyd 's IDEs seem to not like using `.?` TS and flags it as an errors (prob a config issue on DCR). Also I assume if isPaidContent is undefined it should be the same as if it is false?

File: src/web/browser/App.tsx [hunk 2]
```diff
@@ -148,7 +154,7 @@ const App = ({ CAPI, NAV }: Props) => {
                 ))}
 
             <Portal root="cmp">
-                <CMP cmpUi={CAPI.config.switches.cmpUi} />
+                <CMP cmpUi={CAPI.config.cmpUi} />
```
Comment (oliverlloyd) [inline src/web/browser/App.tsx]: We might need to check if there is a semantic purpose to the switches property

File: src/model/window-guardian.ts [hunk 3]
```diff
@@ -66,18 +66,67 @@ const makeWindowGuardianConfig = (
     } as WindowGuardianConfig;
 };
 
+export const makeGuardianBrowserCAPI = (CAPI: CAPIType): CAPIBrowserType => ({
+    config: {
+        isDev: process.env.NODE_ENV !== 'production',
+        ajaxUrl: CAPI.config.ajaxUrl,
+        shortUrlId: CAPI.config.shortUrlId,
```
Comment (oliverlloyd) [inline src/model/window-guardian.ts]: So I don't think this needs to happen in this PR but just so we don't forget: things like this (`shortUrlId`) aren't really 'config', they're properties of the article. Now that we have control of this type, it'd be great to have another PR later to review this structure and rationalise things

File: src/model/window-guardian.ts [hunk 4]
```diff
@@ -66,18 +66,67 @@ const makeWindowGuardianConfig = (
...
+        pageId: CAPI.config.pageId,
+        isPaidContent: !!CAPI.config.isPaidContent,
+        showRelatedContent: CAPI.config.showRelatedContent,
+        keywordIds: CAPI.config.keywordIds,
+        ampIframeUrl: CAPI.config.ampIframeUrl,
+        // commercialBundleUrl is only used in SSR
+        // commercialBundleUrl: CAPI.config.commercialBundleUrl,
```
Comment (oliverlloyd) [inline src/model/window-guardian.ts]: If they're not used can we delete them?

File: src/model/window-guardian.ts [hunk 5]
```diff
@@ -66,18 +66,67 @@ const makeWindowGuardianConfig = (
...
+    blocks: CAPI.blocks,
+    editionId: CAPI.editionId,
+    pillar: CAPI.pillar,
+    contentType: CAPI.contentType,
+    sectionName: CAPI.sectionName,
+    shouldHideReaderRevenue: CAPI.shouldHideReaderRevenue,
+    pageType: {
+        isMinuteArticle: CAPI.pageType.isMinuteArticle,
+        isPaidContent: CAPI.pageType.isPaidContent,
+    },
```
Comment (oliverlloyd) [inline src/model/window-guardian.ts]: I've always found `pageType` to be a needless abstraction (from DCR's point of view) so maybe later when we review this structure we can lose this?

File: src/model/window-guardian.ts [hunk 6]
```diff
@@ -66,18 +66,67 @@ const makeWindowGuardianConfig = (
...
-        data: DCRDocumentData;
+        data: DCRBrowserDocumentData;
         cssIDs: string[];
     };
 
-    // The 'config' attribute is derived from DCRDocumentData and contains
+    // The 'config' attribute is derived from DCRServerDocumentData and contains
     // all the data that, for legacy reasons, for instance compatibility
     // with the frontend commercial stack, or other scripts, we want to find
     // at window.guardian.config
```
[the remark this answers is not in the document]
Comment (liywjl) [reply src/model/window-guardian.ts]: yes this remains true, we dnt touch `window.guardian.config` bec it is used by commercial

File: src/model/window-guardian.ts [hunk 7]
```diff
@@ -92,13 +141,16 @@ export interface WindowGuardian {
 }
 
 export const makeWindowGuardian = (
-    dcrDocumentData: DCRDocumentData,
+    dcrDocumentData: DCRServerDocumentData,
```
[the remark this answers is not in the document]
Comment (liywjl) [reply src/model/window-guardian.ts]: `makeWindowGuardian` is used to create the window object, so I dnt quite follow

File: src/web/browser/App.tsx (hunk 2 above)
Comment (liywjl) [reply src/web/browser/App.tsx]: switches appear just to be boolean flags, so  I am not sure why they need thier own sub section in configs

File: src/model/window-guardian.ts (hunk 5 above)
Comment (liywjl) [reply src/model/window-guardian.ts]: agreeded

**Comment by oliverlloyd** [comment]
I checked the Lighthouse figures here and saw a 3 point increase. This was based on 10 runs locally, without ads:


  | After | Before
-- | -- | --
  | 72 | 65
  | 69 | 65
  | 70 | 71
  | 72 | 69
  | 71 | 70
  | 71 | 66
  | 71 | 68
  | 73 | 72
  | 70 | 69
  | 73 | 70
Average: | 71.2 | 68.5
  |   |  
Difference: | 2.7 |

**Comment by liywjl** [comment]
> I checked the Lighthouse figures here and saw a 3 point increase. This was based on 10 runs locally, without ads:
> 
>  	After	Before
>  	72	65
>  	69	65
>  	70	71
>  	72	69
>  	71	70
>  	71	66
>  	71	68
>  	73	72
>  	70	69
>  	73	70
> Average:	71.2	68.5
>  	 	 
> Difference:	2.7

not bad, I assume this will rise further with the removal of blocks from the frontend CAPI

**Comment by oliverlloyd** [comment]
> not bad, I assume this will rise further with the removal of blocks from the frontend CAPI

3 points is loads considering this is a local run. I'm not sure dropping `blocks` will do much. It's not that big compared to the rest of the object. Analysing this change I can see the size of the data is reduced by about 50% with this change. The blocks is only a small part of that.

**Comment by oliverlloyd** [comment]
One possible big win remaining is `NAV` - there's a lot of data in there we need but it's nonetheless quite large and I suspect could be refactored to be smaller

**Pull request merged**